keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+* Additional node labels can be made available via downward API using admission webhooks that mutate `pods/binding`.
+
+### Non-Goals
+
+* Exposing non-standard node labels by default
+* Enhnacements to topology-aware scheduling
+* Changes to standard topology labels in Kubernetes
+* Changes to downward API
+
+## Proposal
+
+Retreival of topology information will be achieved through the existing mechanism of passing down Pod labels using Downward API.
+A new admission plugin, `PodTopologyLabels`, will be added to kube-apiserver. When enabled, it will mutate the `pods/binding` subresource,
+adding topology labels that match those of the target Node. The Binding REST implementation will be updated to copy these labels
+from the `pods/binding` subresource to the assigned Pod's labels.
+
+Using the downward API to retrieve topology information will look similar to the following:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  generateName: downwardapi-example-
+spec:
+  containers:
+    - name: container
+      image: registry.k8s.io/busybox
+      command: ["sh", "-c"]
+      args:
+      - while true; do
+          if [[ -e /etc/podinfo/zone ]]; then
+            echo -en '\n\n'; cat /etc/podinfo/zone; fi;
+          sleep 5;
+        done;
+      volumeMounts:
+        - name: podinfo
+          mountPath: /etc/podinfo
+  volumes:
+    - name: podinfo
+      downwardAPI:
+        items:
+          - path: "zone"
+            fieldRef:
+              fieldPath: metadata.labels['topology.kubernetes.io/zone']
+```
+
+### User Stories
+
+* As an ML engineer, I want to optimize GPU-to-GPU communication during training which requires topology-awareness in my training code.
+* As a database developer, I want to leverage topology information to improve fault tolerance of sharded databases on Kubernetes.
+* As a developer of a Kubernetes CNI plugin, I want to pass topology information down to the CNI plugin to optimize data paths for container networks.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+* Scope creep. Allowing additional node information or node label info could
+create security issues. This is mitigated by limiting the node labels
+to strictly those that are standardized through KEP-1659.
+
+* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
+standard topology labels are available to Pods.
+
+* Stale data. Information obtained through node labels is like information
+attained through a configmap or secret mounted to a pod, being passed on
+creation but not guaranteed to be immutable and thus should be treated as so.
+
+
+## Design Details
+
+* A built-in Kubernetes admission plugin, `PodTopologyLabels` will be introduced in kube-apiserver
+* The `PodTopologyLabels` admission plugin is responsible for mutating `pods/binding` subresource, adding topology labels matching the target Node.
+* `PodTopologyLabels` admission will not override any existing topology labels on Pods.
```

Comment:
does this one run after all customer's mutating hooks?

Reply:
It will depend on re-invocation policy https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy